src/transformers/models/fastspeech2_conformer/modeling_fastspeech2_conformer.py

```diff
@@ -0,0 +1,1781 @@
+        query = self.linear_q(query).view(batch_size, -1, self.num_attention_heads, self.dim_key)
+        key = self.linear_k(key).view(batch_size, -1, self.num_attention_heads, self.dim_key)
+        value = self.linear_v(value).view(batch_size, -1, self.num_attention_heads, self.dim_key)
+
+        query = query.transpose(1, 2)
+        key = key.transpose(1, 2)
+        value = value.transpose(1, 2)
+
+        return query, key, value
+
+    def forward_attention(self, value, scores, mask):
+        """
+        Compute attention context vector.
+
+        Args:
+            value (torch.Tensor): Transformed value (batch, num_head, time2, dim_key).
+            scores (torch.Tensor): Attention score (batch, num_head, time1, time2).
+            mask (torch.Tensor): Mask (batch, 1, time2) or (batch, time1, time2).
+
+        Returns:
+            torch.Tensor: Transformed value (batch, time1, d_model)
+                weighted by the attention score (batch, time1, time2).
+        """
+        batch_size = value.size(0)
+        if mask is not None:
+            # (batch_size, 1, *, time2)
+            mask = mask.unsqueeze(1).eq(0)
+            min_value = float(numpy.finfo(torch.tensor(0, dtype=scores.dtype).numpy().dtype).min)
+            scores = scores.masked_fill(mask, min_value)
+            # (batch_size, head, time1, time2)
+            attn = torch.softmax(scores, dim=-1).masked_fill(mask, 0.0)
+        else:
+            # (batch_size, head, time1, time2)
+            attn = torch.softmax(scores, dim=-1)
+
+        dropout_applied_attn_scores = self.dropout(attn)
+        # (batch_size, head, time1, dim_key)
+        attention_weighted_values = torch.matmul(dropout_applied_attn_scores, value)
+        # (batch_size, time1, d_model)
+        attention_weighted_values = (
+            attention_weighted_values.transpose(1, 2)
+            .contiguous()
+            .view(batch_size, -1, self.num_attention_heads * self.dim_key)
+        )
+
+        attention_output = self.linear_out(attention_weighted_values)
+
+        return attention_output, attn
+
+    def rel_shift(self, pos_tensor):
+        """
+        Args:
+        Compute relative positional encoding.
+            pos_tensor (torch.Tensor): Input tensor (batch_size, head, time1, 2*time1-1). time1 means the length of
+            query vector.
+        Returns:
+            torch.Tensor: Output tensor.
+        """
+        zero_pad = torch.zeros((*pos_tensor.size()[:3], 1), device=pos_tensor.device, dtype=pos_tensor.dtype)
+        pos_tensor_padded = torch.cat([zero_pad, pos_tensor], dim=-1)
+
+        pos_tensor_padded = pos_tensor_padded.view(*pos_tensor.size()[:2], pos_tensor.size(3) + 1, pos_tensor.size(2))
+        # only keep the positions from 0 to time2
+        pos_tensor = pos_tensor_padded[:, :, 1:].view_as(pos_tensor)[:, :, :, : pos_tensor.size(-1) // 2 + 1]
+
+        return pos_tensor
+
+    def forward(self, query, key, value, pos_emb, mask):
```

Comment:
Did a refactor to bring it more into alignment with the ones FS2 has, but note that FS2 does not use ones like `past_key_value` since its non auto-regressive

Reply:
Perfect - that's great thanks